### DISCUSSION TITLE
Does this library support interop with pycrypto?

### DISCUSSION THREAD

**Discussion by ColtonIdle** [whyoleg/cryptography-kotlin#41] [Q&A] [answered]
Just getting started in this field but for my job i have to provide interop (both ways) with a python library.

https://pycryptodome.readthedocs.io/en/latest/src/signature/dsa.html

so basically im trying to replicate these two functions in kotlin

```
A sender can use a private key (loaded from a file) to sign a message:

from Crypto.Hash import SHA256
from Crypto.PublicKey import ECC
from Crypto.Signature import DSS

message = b'I give my permission to order #4355'
key = ECC.import_key(open('privkey.der').read())
h = SHA256.new(message)
signer = DSS.new(key, 'fips-186-3')
signature = signer.sign(h)
```

and

```
The receiver can use the matching public key to verify authenticity of the received message:

from Crypto.Hash import SHA256
from Crypto.PublicKey import ECC
from Crypto.Signature import DSS

key = ECC.import_key(open('pubkey.der').read())
h = SHA256.new(received_message)
verifier = DSS.new(key, 'fips-186-3')
try:
    verifier.verify(h, signature)
    print("The message is authentic.")
except ValueError:
    print("The message is not authentic.")
```

but i dont see any mention of fips-186-3 in this library/sample code so im not sure if that fips-186-3 is just a standard and i dont need to actually declare that in my kotlin code anywhere?

**Comment by whyoleg** [comment, marked as answer]
Hey! Yeah, there is no need to provide `fips-186-3`, this is the default behaviour now. So everything should work fine.
The main difference with this `python` library is that you don't need to hash data yourself, but it will be done under the hood.

So to mimic your code here, it will be something like this:
```kotlin
val ecdsa = CryptographyProvider.Default.get(ECDSA)
val signature =
    ecdsa
        .privateKeyDecoder(EC.Curve.P521)
        .decodeFrom(EC.PrivateKey.Format.DER, TODO("read bytes from 'privkey.der'"))
        .signatureGenerator(SHA256, ECDSA.SignatureFormat.RAW) // python lib used RAW format by default
        .generateSignature("I give my permission to order #4355".encodeToByteArray())

val isValid =
    ecdsa
        .publicKeyDecoder(EC.Curve.P521)
        .decodeFrom(EC.PublicKey.Format.DER, TODO("read bytes from 'pubkey.der'"))
        .signatureVerifier(SHA256, ECDSA.SignatureFormat.RAW)
        .verifySignature("I give my permission to order #4355".encodeToByteArray(), signature)
check(isValid) { "Signature is invalid" }
```

The only thing that is not supported is automatic decoding of curves - you need to provide it explicitly.

Hope this helps!

**Reply by ColtonIdle** [reply]
Oh heck yeah! This is insanely helpful as I've been going insane trying to match that same logic.

I noticed in the docs that this _might_ not be supported on iOS. Is that right?

I was able to get it working with CryptoKit/Security so it's not the biggest deal. but would be nice to have a single interface for it.

FWIW, this was my code that worked (to decode at least. haven't done encoding/signing)

let publicKey = try! P256.Signing.PublicKey(pemRepresentation: pemString)

let signature = try! P256.Signing.ECDSASignature(rawRepresentation: Data(hex: "hex representation here")!)

let isValid = publicKey.isValidSignature(signature, for: "put the bytes here".data(using: .utf8)!)
print("Is valid: \(isValid)")

**Reply by whyoleg** [reply]
It's supported only with `openssl` provider on iOS in 0.3.1.
But in the next release, there will be support also in `apple` provider, which is also available for iOS.

**Reply by ColtonIdle** [reply]
Interesting! This is all still new to me. so it seems like I can use this with iOS! Awesome. I just need to use the openssl provider on apple as well? Hm. I'm new to kmp, let me see if I can get my dependencies looking right

maybe something like this should work

```

iosMain.dependencies {
    implementation("dev.whyoleg.cryptography:cryptography-provider-openssl:0.3.1")
}

androidMain.dependencies {
    implementation("dev.whyoleg.cryptography:cryptography-provider-jdk:0.3.1")
}

commonMain.dependencies {.   
implementation("dev.whyoleg.cryptography:cryptography-core:0.3.1")
}
```

**Reply by whyoleg** [reply]
Yes, you got it right!

here is an example: https://github.com/whyoleg/cryptography-kotlin/issues/27#issuecomment-2245797420

```kotlin
kotlin {
    sourceSets {
        commonMain.dependencies {
            implementation(project.dependencies.platform("dev.whyoleg.cryptography:cryptography-bom:0.3.1"))
            implementation("dev.whyoleg.cryptography:cryptography-core")
        }
        androidMain.dependencies {
            implementation("dev.whyoleg.cryptography:cryptography-provider-jdk")
        }
        iosMain.dependencies {
            implementation("dev.whyoleg.cryptography:cryptography-provider-openssl3-prebuilt")
            // or `apple` provider
            // implementation("dev.whyoleg.cryptography:cryptography-provider-apple")
        }
    }
}
```

P.S. I will definitely add this to the docs/readme in upcoming release...

**Reply by ColtonIdle** [reply]
Ah nice. That is exactly what i was looking for! updating the readme will be excellent. cheers
